graph/coloring/coloring.go

```diff
@@ -0,0 +1,756 @@
+// Copyright ©2021 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// See Lewis, A Guide to Graph Colouring; Algorithms and Applications
+// doi:10.1007/978-3-319-25730-3 for significant discussion of approaches.
+
+package coloring
+
+import (
+	"context"
+	"errors"
+	"sort"
+
+	"golang.org/x/exp/rand"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/internal/set"
+	"gonum.org/v1/gonum/graph/iterator"
+	"gonum.org/v1/gonum/graph/topo"
+)
+
+// ErrInvalidPartialColoring is returned when a partial coloring
+// is provided for a graph with inadmissible color assignments.
+var ErrInvalidPartialColoring = errors.New("coloring: invalid partial coloring")
+
+// Sets returns the mapping from colors to the IDs of set of nodes of each
+// color. Each set of node IDs is sorted by ascending value.
+func Sets(colors map[int64]int) map[int][]int64 {
+	sets := make(map[int][]int64)
+	for id, c := range colors {
+		sets[c] = append(sets[c], id)
+	}
+	for _, s := range sets {
+		sort.Slice(s, func(i, j int) bool { return s[i] < s[j] })
+	}
+	return sets
+}
+
+// Dsatur returns an approximate minimal chromatic number of g and a
+// corresponding vertex coloring using the heuristic Dsatur coloring algorithm.
+// If a partial coloring is provided the coloring will be consistent with
+// that partial coloring if possible. Otherwise Dsatur will return
+// ErrInvalidPartialColoring.
+// See Brélaz doi:10.1145/359094.359101 for details of the algorithm.
+func Dsatur(g graph.Undirected, partial map[int64]int) (k int, colors map[int64]int, err error) {
+	nodes := g.Nodes()
+	n := nodes.Len()
+	if n == 0 {
+		return
+	}
+	partial, ok := isValid(partial, g)
+	if !ok {
+		return -1, nil, ErrInvalidPartialColoring
+	}
+	order := sortBySaturationDegree(nodes, g, partial)
+	order.heuristic = order.dsatur
+	k, colors = greedyColoringOf(g, order, order.colors)
+	return k, colors, nil
+}
+
+// DsaturExact returns the exact minimal chromatic number of g and a
+// corresponding vertex coloring using the branch-and-bound Dsatur coloring
+// algorithm of Brélaz. If the provided context is cancelled or times out
+// before completion, the context's reason for termination will be returned
+// along with a potentially sub-optimal chromatic number and coloring.
+// See Brélaz doi:10.1145/359094.359101 for details of the algorithm.
+func DsaturExact(ctx context.Context, g graph.Undirected) (k int, colors map[int64]int, err error) {
+	// This is implemented essentially as described in algorithm 1 of
+	// doi:10.1002/net.21716 with the exception that we obtain a
```

Comment:
For future, it can be found here https://docs.switzernet.com/people/emin-gabrielyan/060708-thesis-ref/papers/Brelaz79.pdf (this has been [saved in the wayback machine](https://web.archive.org/web/20210615023929/https://docs.switzernet.com/people/emin-gabrielyan/060708-thesis-ref/papers/Brelaz79.pdf) in case this page disappears).